### PR TITLE
Add str.as_str() for easy Deref to string slices

### PR DESCRIPTION
**Pull request by kornelski**
Working with `Box<str>` is cumbersome, because in places like `iter.filter()` it can end up being `&Box<str>` or even `&&Box<str>`, and such type doesn't always get auto-dereferenced as expected. 

Dereferencing such box to `&str` requires ugly syntax like `&**boxed_str` or `&***boxed_str`, with the exact amount of `*`s. 

`Box<str>` is [not easily comparable with other string types](https://github.com/rust-lang/rust/pull/129852) via `PartialEq`. `Box<str>` won't work for lookups in types like `HashSet<String>`, because `Borrow<String>` won't take types like `&Box<str>`. OTOH `set.contains(s.as_str())` works nicely regardless of levels of indirection.

`String` has a simple solution for this: the `as_str()` method, and `Box<str>` should too.

### CODE REVIEW

**Comment by tgross35** [comment]
r? libs-api

**Comment by Amanieu** [comment]
We discussed this in the @rust-lang/libs-api meeting. We agree that such a method is useful, however it would be better to place this method on `str` directly, which allows it work for other smart pointer types such as `Arc` or `Rc`. Additionally, since this is a new API it must be marked as unstable and a tracking issue needs to be created for it.

**Comment by GrigorenkoPV** [comment]
I've created a tracking issue: #130366

**Comment by joshtriplett** [comment]
r=me once this passes CI.

File: library/std/src/lib.rs [hunk 1]
```diff
@@ -351,6 +351,7 @@
 #![feature(slice_ptr_get)]
 #![feature(slice_range)]
 #![feature(std_internals)]
+#![feature(str_as_str)]
```
Comment (cuviper) [inline library/std/src/lib.rs:354]: Why is this needed? You didn't make any other changes in `std` -- did the new method break something?

File: library/alloc/src/lib.rs [hunk 2]
```diff
@@ -147,6 +147,7 @@
 #![feature(slice_ptr_get)]
 #![feature(slice_range)]
 #![feature(std_internals)]
+#![feature(str_as_str)]
```
Comment (cuviper) [inline library/alloc/src/lib.rs:150]: Assuming this is only for the new tests, it should be added to the earlier `cfg_attr(test, feature(...))` attribute instead.

File: library/std/src/lib.rs (hunk 1 above)
Comment (kornelski) [reply library/std/src/lib.rs:354]: It's been used by a test for an error message specifically about lack of `str.as_str()` support. I've deleted that check.

File: library/alloc/src/lib.rs (hunk 2 above)
Comment (kornelski) [reply library/alloc/src/lib.rs]: Done

**Review by dtolnay — APPROVED** [review]
